src/main/java/com/example/onlinestore/service/impl/UserServiceImpl.java

```diff
@@ -190,4 +194,17 @@ public User getUserByToken(String token) {
             return null;
         }
     }
+
+    /**
+     * 记录登录失败次数，如果超过阈值可以用于后续风控。
+     */
+    private void recordFailedLogin(String username) {
+        String key = "login:fail";
+        Long cnt = redisTemplate.opsForValue().increment(key);
+        if (cnt != null && cnt == 1) {
+            // 设置过期时间
+            redisTemplate.expire(key, 1, TimeUnit.DAYS);
+        }
+        logger.debug("记录失败登录次数 {} -> {}", username, cnt);
+    }
```

Comment:
方法 `recordFailedLogin` 的计数键未区分用户。

1. **[高] 风险**: 使用全局键会使单一攻击或高频错误影响所有用户的风控判断；建议按用户名维度计数并保留原有 TTL。

```suggestion
        String key = "login:fail:" + username;
        Long cnt = redisTemplate.opsForValue().increment(key);
        if (cnt != null && cnt == 1) {
            redisTemplate.expire(key, 1, TimeUnit.DAYS);
        }
        logger.debug("记录失败登录次数 {} -> {}", username, cnt);
```


---
*Powered by Qoder*